libkbfs/tlf_journal_test.go

```diff
@@ -0,0 +1,227 @@
+
+func TestTLFJournalBasic(t *testing.T) {
+	tempdir, config, ctx, cancel, tlfJournal, delegate :=
+		setupTLFJournalTest(t)
+	defer teardownTLFJournalTest(
+		t, ctx, cancel, tlfJournal, delegate, tempdir, config)
+
+	putBlock(ctx, t, config, tlfJournal)
+
+	// Wait for it to be processed.
+
+	delegate.requireNextState(ctx, t, bwBusy)
+	delegate.requireNextState(ctx, t, bwIdle)
+}
+
+func TestTLFJournalPauseResume(t *testing.T) {
+	tempdir, config, ctx, cancel, tlfJournal, delegate :=
+		setupTLFJournalTest(t)
+	defer teardownTLFJournalTest(
+		t, ctx, cancel, tlfJournal, delegate, tempdir, config)
+
+	tlfJournal.pauseBackgroundWork()
+	delegate.requireNextState(ctx, t, bwPaused)
+
+	putBlock(ctx, t, config, tlfJournal)
+
+	// Unpause and wait for it to be processed.
+
+	tlfJournal.resumeBackgroundWork()
+	delegate.requireNextState(ctx, t, bwIdle)
+	delegate.requireNextState(ctx, t, bwBusy)
+	delegate.requireNextState(ctx, t, bwIdle)
+}
+
+func TestTLFJournalPauseShutdown(t *testing.T) {
+	tempdir, config, ctx, cancel, tlfJournal, delegate :=
+		setupTLFJournalTest(t)
+	defer teardownTLFJournalTest(
+		t, ctx, cancel, tlfJournal, delegate, tempdir, config)
+
+	tlfJournal.pauseBackgroundWork()
+	delegate.requireNextState(ctx, t, bwPaused)
+
+	putBlock(ctx, t, config, tlfJournal)
+
+	// Should still be able to shut down while paused.
+}
+
+type hangingBlockServer struct {
+	BlockServer
+	// Closed on put.
+	onPutCh chan struct{}
+}
+
+func (bs hangingBlockServer) Put(
+	ctx context.Context, tlfID TlfID, id BlockID, context BlockContext,
+	buf []byte, serverHalf BlockCryptKeyServerHalf) error {
+	close(bs.onPutCh)
+	// Hang until the context is cancelled.
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+func (bs hangingBlockServer) waitForPut(ctx context.Context, t *testing.T) {
+	select {
+	case <-bs.onPutCh:
+	case <-ctx.Done():
+		require.FailNow(t, ctx.Err().Error())
+	}
+}
+
+func TestTLFJournalBusyPause(t *testing.T) {
+	tempdir, config, ctx, cancel, tlfJournal, delegate :=
+		setupTLFJournalTest(t)
+	defer teardownTLFJournalTest(
+		t, ctx, cancel, tlfJournal, delegate, tempdir, config)
+
+	bs := hangingBlockServer{tlfJournal.delegateBlockServer,
+		make(chan struct{})}
+	tlfJournal.delegateBlockServer = bs
+
+	putBlock(ctx, t, config, tlfJournal)
+
+	bs.waitForPut(ctx, t)
+	delegate.requireNextState(ctx, t, bwBusy)
+
+	// Should still be able to pause while busy.
+
+	tlfJournal.pauseBackgroundWork()
+	delegate.requireNextState(ctx, t, bwPaused)
+}
+
+func TestTLFJournalBusyShutdown(t *testing.T) {
+	tempdir, config, ctx, cancel, tlfJournal, delegate :=
+		setupTLFJournalTest(t)
+	defer teardownTLFJournalTest(
+		t, ctx, cancel, tlfJournal, delegate, tempdir, config)
+
+	bs := hangingBlockServer{tlfJournal.delegateBlockServer,
+		make(chan struct{})}
+	tlfJournal.delegateBlockServer = bs
+
+	putBlock(ctx, t, config, tlfJournal)
+
+	bs.waitForPut(ctx, t)
+	delegate.requireNextState(ctx, t, bwBusy)
+
+	// Should still be able to shut down while busy.
```

Comment:
It won't, I think -- the teardown process calls cancel(), which cancels the main test context, which causes the hanging block/md server to stop hanging and return.

I wish there were a way to check goroutine counts in tests, but alas...